### ISSUE TITLE
Abstract away cancellable and repeatable promise listening.

### ISSUE THREAD

**Issue by trusktr** [trusktr/infamous#150]
We're currently making logic that depends on promises being cancellable in our own way, for example:

https://github.com/trusktr/infamous/blob/96e5a4f94083729bf1788ce77254628aafbc99ed/src/core/ImperativeBase.js#L191-L208

In that example, a `mountcancel` error is a way to cancel logic from firing. When `mountcancel` is received by the try-catch-finally block, the logic that follows after that block is effectively cancelled if the promise was cancelled.

We're currently doing this only in a couple of places in the code base, but in reality there's more places where this needs to be done otherwise logic that should have been cancelled will still fire and cause unexpected bugs (at some point, we're going to get weird bug reports). We need to ensure that all async logic in our code base is cancelled based on certain condition. F.e. if an element has logic to be fired after having been mounted (i.e. after mountPromise resolves), then if the element is removed from a tree an disposed of, then the mountPromise will not ever be resolved, so we need to clean that up. Or for example, if an element was connected and subsequently has logic waiting on mountPromise, but then the element is moved to a different tree, we need to cancel the original logic and queue new logic. One way to handle this is to hide any promises as much as possible, and rather than end users waiting for promises to resolve, we can give them life cycle methods like mountedCallback and unmountedCallback.

A couple ideas how we can make this cleaner without having to re-use that try-catch-finally boilerplate code everywhere:

- Abstract the try-catch-finally technique of cancelling promises into a reusable form, perhaps a function.
- Perhaps use async-csp to create while loops that area easily `break`'d when a cancellation flag is set.
  - For things like mount/unmount, we might have this while loop call mountedCallback and unmountedCallback, then other code just uses the lifecycle methods without re-using the cancellation logic.

**Comment by trusktr** [comment]
This is no longer relevant, because we removed much of the `async` code and replaced it with simpler, easier-to-understand sync code.
